### PR TITLE
Retrieve quiz objects

### PR DESCRIPTION
**Pull request by nononovia**

### CODE REVIEW

File: quiz_onject_retrieval.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from scripts import quiz_object_retrieval
```
Comment (Opeyem1a) [inline quiz_onject_retrieval.py]: Rename this file

File: scripts/quiz_retrieval/quiz_retrieval.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        else:
+            return super().default(z)
+
+
+def quiz_object_retrieval():
+    canvas_api = CanvasAPI()
+    output_dir = OUTPUT_DIR
+    quizzes = canvas_api.get_quizzes()
+    for quiz in quizzes:
+        output_path = setup_submissions_filepath(quiz, output_dir, "quiz_objects", "quiz_object" )
```
Comment (Opeyem1a) [inline scripts/quiz_retrieval/quiz_retrieval.py:24]: `"quiz_objects"` -> `"quiz"`
`"quiz_object"` -> `"quiz_id_"`

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: scripts/quiz_retrieval/quiz_retrieval.py (hunk 2 above)
Comment (nononovia) [reply scripts/quiz_retrieval/quiz_retrieval.py]: I don't see why this naming scheme is less confusing... Plus Bowen already ran the scripts using this naming scheme so No point changing file name, can still change the directory name I guess.

**Review by carson-ricca — CHANGES_REQUESTED** [review]
Looks like there are conflicts that should be resolved first.

File: scripts/quiz_retrieval/quiz_retrieval.py (hunk 2 above)
Comment (Opeyem1a) [reply scripts/quiz_retrieval/quiz_retrieval.py]: I made this comment in our meeting before we ran them so dw now, also it was to (a) make the directory more visually distinct from the file prefix and, more importantly, (b) make it so the filename describes what the number actually means (i.e. which field in the object the number represents)

File: scripts/quiz_retrieval/quiz_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                with open(quiz_object_path, 'r') as f:
+                    file = f.read()
+                    json_file = json.loads(file)
+                    if json_file:
+                        return json_file[0]['title']
+                    else:
+                        return quiz_id + " can't retrieve quiz name"
+
+
+def get_quiz_object(quiz_id):
```
Comment (Opeyem1a) [inline scripts/quiz_retrieval/quiz_retrieval.py:62]: Why does this function exist? We can just directly concatenate the full path with the information we already know and at best, this function can just check if a file exists at that location and raise an error if not.

There's no need to loop through every file looking for a file when we could just directly find the filepath it should be and verify

File: util/canvas_api.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import os
+import datetime
+import json
+
+from util import Encoder, mkdir_if_not_exists
```
Comment (Opeyem1a) [inline util/canvas_api.py:5]: I don't think Encoder is needed in this file

File: scripts/quiz_retrieval/quiz_retrieval.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    """
+    with open(output_filepath, 'w') as f:
+        json_quiz = []
+        quiz_dict = quiz.__dict__
+        quiz_dict.pop('_requester')
+        json_quiz.append(json.dumps(quiz_dict, indent=4, cls=DateTimeEncoder))
+        f.write('[' + ','.join(json_quiz) + ']')
+
+
+def get_quiz_name(quiz_id):
```
Comment (Opeyem1a) [inline scripts/quiz_retrieval/quiz_retrieval.py:40]: Same with the comment below, why not just immediately concatenate together the full path to what the file's name should be? if `quiz_id=123456`, I can immediately look for `'.../quiz_objects/quiz_object_'+quiz_id`, then raise an error if it doesn't exist or I couldn't read it

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: scripts/canvas_submission_retrieval/__init__.py [hunk 6]
```diff
@@ -1 +1,3 @@
-from .canvas_submission_retrieval import canvas_submission_retrieval
+from .canvas_submission_retrieval import canvas_submission_retrieval, setup_submissions_filepath
+
+
```
Comment (carson-ricca) [inline scripts/canvas_submission_retrieval/__init__.py:3]: Too many blank lines?

File: submission_retrieval.py [hunk 7]
```diff
@@ -18,3 +19,5 @@
         canvas_course_id = get_default_course_id()
 
     canvas_submission_retrieval(EXPORT_DIR, encoder, canvas_course_id)
+    quiz_retrieval()
+
```
Comment (carson-ricca) [inline submission_retrieval.py:23]: I think this was removed in a different PR

**Review by carson-ricca — CHANGES_REQUESTED** [review]
I think some of the changes would require a merge with main after Mat's is merged in?

File: requirements.txt [hunk 8]
```diff
@@ -1,10 +1,12 @@
 canvasapi==2.2.0
 pandas==1.3.3
 eel==0.14.0
+numpy==1.21.2
+matplotlib==3.4.3
+seaborn==0.11.2
 scikit-learn==1.0.0
+matplotlib==3.4.2
```
Comment (carson-ricca) [inline requirements.txt:8]: shows up twice

File: scripts/course_performace_analysis/course_performace_analysis.py [hunk 9]
```diff
@@ -191,12 +204,24 @@ def course_performance_analysis(GRADEBOOK_PATH, QUIZSCOREJSON_PATH):
...
+    plt.close()
+    final_score = final_score
+    first_score = df_student_grade_first_attempt[FIRST_ATTEMPT_FINAL_SCORE]
+    difference_score = final_score - first_score
+    wilcoxon = scipy.stats.wilcoxon(final_score, first_score)
+    print(wilcoxon)
 
-if __name__ == '__main__':
-    quiz_path = os.path.join(PARENT_PATH, 'quizzes')
-    course_performance_analysis('grade_book.csv', QUIZ_PATH)
```
Comment (carson-ricca) [inline scripts/course_performace_analysis/course_performace_analysis.py]: Is it supposed to call the function here still?

**Review by carson-ricca — CHANGES_REQUESTED** [review]

**Review by carson-ricca — APPROVED** [review]

**Pull request closed without merging**